contracts/proxies/CurrencyConvertor.sol

```diff
@@ -81,36 +93,27 @@ contract CurrencyConvertor {
     external
     returns (uint256)
   {
+    address self = address(this);
 
-    // Send fromToken to the ExchangeWrapper.
+    // Send fromToken to this contract.
     IERC20(tokenFrom).safeTransferFrom(
       msg.sender,
-      exchangeWrapper,
+      self,
       tokenFromAmount
     );
 
-    address self = address(this);
-
     // Convert fromToken to toToken on the ExchangeWrapper.
+    IERC20(tokenFrom).safeApprove(exchangeWrapper, type(uint256).max);
```

Comment:
I'm worried about the gas costs here. On every deposit to Starkware, we will be doing 2 calls to `safeApprove` (currency convertor approving `makerToken` for exchange wrapper, exchange wrapper approving `makerToken` for 0x).

These approvals could already be sufficient (greater than `tokenFromAmount`), but less than `type(uint256).max` so we would do a needless approval again. This will be rough from a user gas perspective.

Is there any way we can only do calls to `safeApprove` when it's necessary (IE, this contracts allowance for `tokenFrom` is less than `tokenFromAmount`)?

Also, is there any way that we could remove the exchange wrapper contract in the middle, so we don't need to do two additional transfers?

Reply:
the project scope is that we have interchangeable exchange wrappers